handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -303,8 +303,24 @@ public void operationComplete(Future<SslContext> future) throws Exception {
     }
 
     private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
-        this.selection = selection;
-        SslHandler sslHandler = selection.context.newHandler(ctx.alloc());
+        try {
+            this.selection = selection;
+            replaceHandler(ctx, selection.hostname, selection.context);
+        } catch (Throwable cause) {
+            this.selection = EMPTY_SELECTION;
+            ctx.fireExceptionCaught(cause);
```

Comment:
Maybe even fire a custom `SniHandlerException(hostname, context, cause)` to give the user every possible opportunity not to lose track of the SslContext.